### PR TITLE
feature/SOF 7282 1

### PR DESCRIPTION
**Pull request by VsevolodX**
- feat: add commensurate supercells generation builder

### CODE REVIEW

File: src/py/mat3ra/made/tools/build/interface/builders.py [hunk 1]
```diff
@@ -284,3 +286,138 @@ def _update_material_name(
...
+        a = film.lattice.vector_arrays[0][:2]
+        b = film.lattice.vector_arrays[1][:2]
+        commensurate_lattices = self.__generate_commensurate_lattices(a, b, max_search, configuration.twist_angle)
+        commensurate_lattice_pairs = [
+            CommensurateLatticePair(configuration=configuration, **lattice) for lattice in commensurate_lattices
+        ]
+        return commensurate_lattice_pairs
+
+    @staticmethod
+    def __create_matrices(max_search: int):
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/interface/builders.py:325]: create_supercell_matrices

Comment (timurbazhirov) [reply src/py/mat3ra/made/tools/build/interface/builders.py:325]: Can be util

File: src/py/mat3ra/made/tools/build/interface/builders.py [hunk 2]
```diff
@@ -284,3 +286,138 @@ def _update_material_name(
...
+                        # Non-zero area constraint
+                        matrix = np.array([[s11, s12], [s21, s22]])
+                        determinant = np.linalg.det(matrix)
+                        if determinant == 0 or determinant < 0:
+                            continue
+                        matrices.append(matrix)
+        return matrices
+
+    @staticmethod
+    def __solve_angle_from_rotation_matrix(matrix, zero_tolerance=1e-6, round_digits=3):
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/interface/builders.py:340]: Can be util

File: src/py/mat3ra/made/tools/build/interface/builders.py [hunk 3]
```diff
@@ -284,3 +286,120 @@ def _update_material_name(
...
+        matrix_a1a2_inverse = np.linalg.inv(matrix_a1a2)
+
+        solutions = []
+        for index1, matrix1 in enumerate(matrices):
+            for index2, matrix2 in enumerate(matrices[0 : index1 + 1]):
+                matrix2_inverse = np.linalg.inv(matrix2)
+                intermediate_product = matrix2_inverse @ matrix1
+                product = matrix_a1a2_inverse @ intermediate_product @ matrix_a1a2
+                try:
+                    angle = self.__solve_angle_from_rotation_matrix(product)
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/interface/builders.py:374]: We can return None and get back from loop here

File: src/py/mat3ra/made/tools/build/interface/builders.py [hunk 4]
```diff
@@ -284,3 +286,120 @@ def _update_material_name(
...
+                    size_metric = np.linalg.det(matrix_a1a2_inverse @ matrix1 @ matrix_a1a2)
+
+                    if np.abs(angle - target_angle) < self.build_parameters.angle_tolerance:
+                        print(f"Found commensurate lattice with angle {angle} and size metric {size_metric}")
+                        solutions.append(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/interface/builders.py:379]: CommensurateLatticePair should be used

File: src/py/mat3ra/made/tools/build/interface/builders.py [hunk 5]
```diff
@@ -1,6 +1,8 @@
 from typing import Any, List, Optional
 
 import numpy as np
+from mat3ra.made.tools.build.supercell import create_supercell
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/interface/builders.py:4]: May be imported from `...`

File: src/py/mat3ra/made/tools/build/interface/builders.py [hunk 6]
```diff
@@ -284,3 +245,96 @@ def _update_material_name(
     ) -> Material:
         material.name = f"Twisted Nanoribbon Interface ({configuration.twist_angle:.2f}°)"
         return material
+
+
+class CommensurateLatticeInterfaceBuilderParameters(BaseModel):
+    max_search: int = 10
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/interface/builders.py:251]: `max_repetition_int`

Comment (timurbazhirov) [reply src/py/mat3ra/made/tools/build/interface/builders.py:251]: + Docstring

File: src/py/mat3ra/made/tools/build/interface/builders.py [hunk 7]
```diff
@@ -284,3 +245,96 @@ def _update_material_name(
...
+    ) -> List[Material]:
+        interfaces = []
+        for item in items:
+            new_substrate = create_supercell(item.configuration.film, item.matrix1.tolist())
+            new_film = create_supercell(item.configuration.substrate, item.matrix2.tolist())
+            new_film = translate_by_vector(
+                new_film, [0, 0, item.configuration.distance_z], use_cartesian_coordinates=True
+            )
+            try:
+                interface = merge_materials([new_substrate, new_film])
```
[the remark this answers is not in the document]
Comment (VsevolodX) [reply src/py/mat3ra/made/tools/build/interface/builders.py:336]: The try case? lms

File: src/py/mat3ra/made/utils.py [hunk 8]
```diff
@@ -77,6 +77,57 @@ def get_overlapping_coordinates(
...
+    """
+    matrices = []
+    for s11 in range(-max_search, max_search + 1):
+        for s12 in range(-max_search, max_search + 1):
+            for s21 in range(-max_search, max_search + 1):
+                for s22 in range(-max_search, max_search + 1):
+                    # Non-zero area constraint
+                    matrix = np.array([[s11, s12], [s21, s22]])
+                    determinant = np.linalg.det(matrix)
+                    if determinant == 0 or determinant < 0:
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/utils.py:96]: This needs to be explained either docstring or here: `det < 0` is excluded b/c of mirror transformations

File: src/py/mat3ra/made/utils.py [hunk 9]
```diff
@@ -77,6 +77,57 @@ def get_overlapping_coordinates(
...
+                        continue
+                    matrices.append(matrix)
+    return matrices
+
+
+def get_angle_from_rotation_matrix(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/utils.py:102]: get_angle_from_rotation_matrix_2d

File: src/py/mat3ra/made/tools/build/interface/commensurate_lattice_pair.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+from .configuration import TwistedInterfaceConfiguration
+
+
+class CommensurateLatticePair(BaseModel):
+    """
+    Commensurate lattice pair model.
+
+    Attributes:
+        configuration (TwistedInterfaceConfiguration): The configuration of the twisted interface.
+        matrix1 (np.ndarray): The supercell matrix for the first lattice.
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/interface/commensurate_lattice_pair.py:13]: 2d supercell matrix

File: src/py/mat3ra/made/tools/build/interface/builders.py [hunk 11]
```diff
@@ -284,3 +245,111 @@ def _update_material_name(
...
+class CommensurateLatticeInterfaceBuilder(BaseBuilder):
+    _GeneratedItemType: type(CommensurateLatticePair) = CommensurateLatticePair  # type: ignore
+    _ConfigurationType: type(TwistedInterfaceConfiguration) = TwistedInterfaceConfiguration  # type: ignore
+
+    def _generate(self, configuration: _ConfigurationType) -> List[_GeneratedItemType]:
+        film = configuration.film
+        # substrate = configuration.substrate
+        max_search = self.build_parameters.max_repetition_int
+        a = film.lattice.vector_arrays[0][:2]
+        b = film.lattice.vector_arrays[1][:2]
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/interface/builders.py:274]: We should have a getter for this `film.lattice.a`, `film.lattice.b`, `film.lattice.c`

Comment (VsevolodX) [reply src/py/mat3ra/made/tools/build/interface/builders.py:275]: Ok. We're adding this across all the made during cleanup task.

**Review by timurbazhirov — APPROVED** [review]

**Pull request merged**